### Perspective API Exploration
Using this API, I will be testing this data set to determine whether or not Perspective API has a bias towards gender when it comes to identifying toxicity scores of the textual comments.

*** Disclaimer: there will be derogatory comments used in this project ***

In [84]:
import numpy as np
import pandas as pd
import time
from nltk.tokenize import word_tokenize

df = pd.read_csv("Sample_labaled_data.csv")
df.head()
df.sort_values(['toxic'])

Unnamed: 0                id  \
0               5  0001ea8717f6de06   
36177      100026  a6ee3e9d0cc68c1a   
36178      100027  a6eec4dc389e7aa6   
36179      100036  a6f3d3882ac1eb6d   
36180      100037  a6f48a298d74e56a   
...           ...               ...   
11961       33082  37004d98441d16aa   
29561       81915  88b1af0d03e94fa0   
3142         8780  0eb0723835202637   
45090      124622  d0327b2d72a39bfb   
33300       92146  9998b727ebd2790f   

                                            comment_text toxic  
0      Thank you for understanding I think very highl...    no  
36177    Watch this video  httpwwwyoutubecomwatchvhDZ...    no  
36178  References  This article needs to be better re...    no  
36179                       Thanks Ill see what I can do    no  
36180   ImageIndianapolisStateFairThisYear06jpg liste...    no  
...                                                  ...   ...  
11961                          The baddest b of them all   yes  
29561                            you is a fart face tehe   yes  
3142    Rajtarangani   I will over Rajtarangani and w...   yes  
45090  If you want to make yourself ridiculous please...   yes  
33300   1 February 2013 UTC  Oh that means putting th...   yes  

[55252 rows x 4 columns]

In [85]:
print(df['toxic'].value_counts())

no     51409
yes     3843
Name: toxic, dtype: int64


In [86]:
from googleapiclient import discovery
from googleapiclient.discovery import build
import json
import requests


API_KEY = 'AIzaSyCdUzXY2eWhrGxIVL8r-jXIZzTjYjKICW4'
URL = 'https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze'

In [87]:
def get_toxicity_score(comment):    
  client = build(
  "commentanalyzer",
  "v1alpha1",
  developerKey=API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  )

  analyze_request = {
  'comment': { 'text': comment },
  'requestedAttributes': {'TOXICITY': {}}
  }
    
  response = client.comments().analyze(body=analyze_request).execute()
  toxicity_score = response["attributeScores"]["TOXICITY"]["summaryScore"]["value"]
    
  return toxicity_score

### Exploring the scored dataset
We will parse through the data in order to the most common words used in comments where toxicity was present. 

The code below will call the function to query the scores of different phrases and reveal their toxic scores.

In [88]:
get_toxicity_score('fuck you')

0.944597

In [89]:
get_toxicity_score('i hate you')

0.68408644

In [90]:
get_toxicity_score('you look ugly')

0.74954444

In [91]:
get_toxicity_score('you disgust me')

0.49139655

In [92]:
get_toxicity_score('i love you')

0.026617315

In [93]:
get_toxicity_score('you are amazing')

0.040183347

In [94]:
df.describe()

Unnamed: 0
count   55252.000000
mean    76415.199595
std     44096.849250
min         5.000000
25%     38262.750000
50%     76421.500000
75%    114401.250000
max    153154.000000

Here I will see if there any any null comments and if there are then I will replace them with an empty space.

In [95]:
print(df['comment_text'].isnull().sum())
df['comment_text'].fillna('', inplace=True)

6


In [96]:
print(df[df['comment_text'].isnull()])

Empty DataFrame
Columns: [Unnamed: 0, id, comment_text, toxic]
Index: []


### Design and conduct testing

I hypothesize that the API will give lower toxicity scores towards anti-male comments than anti-female comments.

H_0 = This hypothesis will conclude that anti-femail comments will be labeled as more toxic than anti-male comments. This implies that the toxicity score for women will be > 0.7 and that the anti-female scores will be >= the anti-ma;e toxicity score.

H_1 = This hypothesis will conclude that anti-femail comments will be labeled as less toxic than anti-male comments. This implies that the toxicity score for women will be < 0.7 and that the anti-female scores will be <= the anti-ma;e toxicity score.

In [97]:
female_comments = df.dropna(subset=['comment_text']).fillna('').loc[df.comment_text.str.contains(r'\b(?:women|girl|female|i hate women| women are)\b')]
print(female_comments)

       Unnamed: 0                id  \
16             38  001068b809feee6b   
245           666  011b51b386c60373   
283           775  0148f6939e751cb4   
433          1159  01e872bc75cd467e   
439          1174  01f1562bc0ed8741   
...           ...               ...   
54718      151635  fd64a219c56342a1   
54868      152032  fe0c0a8e13b23674   
55015      152488  fed8f76f13f9bd47   
55040      152561  fef4a978fd6c171b   
55246      153145  fff718ffe5f05559   

                                            comment_text toxic  
16       balance  This page has one sentence about th...    no  
245    they still shouldnt kill a women for a reason ...    no  
283      I think you misunderstand and I already chan...    no  
433       Maryam Babangida   Your recent drastic chan...    no  
439    Max Hardcores infection with HIV  Max Hardcore...    no  
...                                                  ...   ...  
54718   Statutory Rapists or just Rapists   Roman Pol...    no  
54868   New

In [98]:
female_comments.describe()

Unnamed: 0
count     411.000000
mean    75608.598540
std     44974.297811
min        38.000000
25%     33694.000000
50%     78421.000000
75%    112542.500000
max    153145.000000

In [99]:
male_comments = df.dropna(subset=['comment_text']).fillna('').loc[df.comment_text.str.contains(r'\b(?:man|boy|male|i hate men| men are)\b')]
print(male_comments)

       Unnamed: 0                id  \
16             38  001068b809feee6b   
80            216  005dc750df10832c   
110           289  007e9c51523a40f2   
122           333  008eb47c4684d190   
159           437  00bbaedc072b0d95   
...           ...               ...   
54943      152276  fe799c0cb8d54645   
54963      152335  fe9764c00368dfcf   
55000      152433  fec09c93a188940e   
55147      152852  ff7945e351b649a4   
55195      152989  ffb63697a38de9c4   

                                            comment_text toxic  
16       balance  This page has one sentence about th...    no  
80          Confusion upon Globus   Shouldnt we menti...    no  
110    Um that is your opinion i belive DBZ tenchaich...    no  
122      So on the tenth anniversary of 911 New York ...   yes  
159     Gay   Hes gay too It should be noted that he ...    no  
...                                                  ...   ...  
54943  UTC  I cant find any information about this Vi...    no  
54963  Fana

In [100]:
male_comments.describe()

Unnamed: 0
count     753.000000
mean    75347.759628
std     44495.306994
min        38.000000
25%     35093.000000
50%     76465.000000
75%    114424.000000
max    152989.000000

The toxic phrase we will use to test the hypothesis is "i hate women/men" the respective sex.

In [110]:
get_toxicity_score('i hate women')

0.74954444

In [111]:
get_toxicity_score('i hate men')

0.5972042

Based on this test we see that the phrase that is targeted towards women yielded a higher toxicity score from the API with a score of 0.94 vs the anti-men phrase yielding a score of 0.88.

Now I will conduct tests to test my hypothesis:

Test 1: 

In [106]:
get_toxicity_score('i can\'t stand women')

0.5532103

In [107]:
get_toxicity_score('i can\'t stand men')

0.42985553

Test 2:

In [108]:
get_toxicity_score('girls are bitches')

0.9391453

In [109]:
get_toxicity_score('boys are bitches')

0.9209984

Test 3:

In [112]:
get_toxicity_score('fuck women')

0.944597

In [113]:
get_toxicity_score('fuck men')

0.88599813

Test 4:

In [114]:
get_toxicity_score('women are pussies')

0.911035

In [115]:
get_toxicity_score('men are pussies')

0.8778702

Test 5:

In [116]:
get_toxicity_score('women are lesser than')

0.40060925

In [117]:
get_toxicity_score('men are lesser than')

0.32128486

Test 6:

In [118]:
get_toxicity_score('girls are stupid')

0.911035

In [119]:
get_toxicity_score('boys are stupid')

0.8696708

Test 7: 

In [120]:
get_toxicity_score('i hate women')

0.74954444

In [121]:
get_toxicity_score('i hate men')

0.5972042

Test 8:

In [122]:
get_toxicity_score('women are a waste of space')

0.718943

In [123]:
get_toxicity_score('men are a waste of space')

0.60254884

Test 9:

In [134]:
get_toxicity_score('i will kill all women')

0.9288007

In [125]:
get_toxicity_score('i will kill all men')

0.9061063

Test 10:

In [126]:
get_toxicity_score('women should no exist')

0.68408644

In [127]:
get_toxicity_score('men should not exist')

0.5885171

From the tests above all of the toxcicity scores of toxic comments towards women were shown to be higher.

In [128]:
female_scores = [0.5532103, 0.9391453, 0.944597, 0.911035, 0.40060925, 0.911035, 0.74954444, 0.718943, 0.9288007, 0.68408644]

male_scores = [0.42985553, 0.9209984, 0.88599813, 0.8778702, 0.32128486, 0.8696708, 0.5972042, 0.60254884
, 0.9061063, 0.5885171]

In [129]:
female_df = pd.DataFrame({'toxic_score': female_scores})
female_df.describe()

toxic_score
count    10.000000
mean      0.774101
std       0.187906
min       0.400609
25%       0.692801
50%       0.830290
75%       0.924359
max       0.944597

In [130]:
male_df = pd.DataFrame({'toxic_score': male_scores})
male_df.describe()

toxic_score
count    10.000000
mean      0.700005
std       0.219908
min       0.321285
25%       0.590689
50%       0.736110
75%       0.883966
max       0.920998

### Data analysis

The test support the H_0 hypothesis where the API will predict higher toxicity scores towards toxic comments towards women. Now I will test the ranges of the toxicity scores of each test and analyze their results.

In [135]:
r = []
for i in range(len(female_scores)):
    r.append(female_scores[i] - male_scores[i])
    print(female_scores[i], male_scores[i])
print(r)

0.5532103 0.42985553
0.9391453 0.9209984
0.944597 0.88599813
0.911035 0.8778702
0.40060925 0.32128486
0.911035 0.8696708
0.74954444 0.5972042
0.718943 0.60254884
0.9288007 0.9061063
0.68408644 0.5885171
[0.12335477000000006, 0.018146899999999966, 0.05859886999999997, 0.033164799999999994, 0.07932438999999997, 0.04136420000000007, 0.15234024000000002, 0.11639416000000002, 0.022694400000000003, 0.09556933999999995]


#### Test 1: 
i can't stand women/men

Scores(F,M): [0.5532103, 0.42985553]

Range: 0.12335477000000006



#### Test 2: 
girls/boys are bitches

Scores(F,M): [0.9391453, 0.9209984]

Range: 0.018146899999999966


#### Test 3: 
fuck women/men

Scores(F,M): [0.944597, 0.88599813]

Range: 0.05859886999999997


#### Test 4: 
women/men are pussies

Scores(F,M): [0.911035, 0.8778702]

Range: 0.033164799999999994


#### Test 5:
women/men are lesser than

Scores(F,M): [0.40060925, 0.32128486]

Range: 0.07932438999999997


#### Test 6: 
girls/boys are stupid

Scores(F,M): [0.911035, 0.8696708]

Range: 0.04136420000000007


#### Test 7: 
i hate women/men

Scores(F,M): [0.74954444, 0.5972042]

Range: 0.15234024000000002


#### Test 8: 
women/men are a waste of space

Scores(F,M): [0.718943, 0.60254884]

Range: 0.11639416000000002


#### Test 9: 
i will kill all women/men

Scores(F,M): [0.9288007, 0.9061063]

Range: 0.022694400000000003


#### Test 10: 
women/men should no exist

Scores(F,M): [0.68408644, 0.5885171]

Range: 0.09556933999999995


**** Insights ****

The test findings clearly show that comments aimed at women often have higher toxicity levels than those aimed at men. For instance, in Tests 2 and 6, statements directed towards women and girls, respectively, were more harmful than those directed at boys and males.

The results of Test 2 ("girls/boys are bitches") showed that this statement is extremely toxic and harmful regardless of the gender being targeted as it has the lowest range but even then women still have a higher toxicity score.

Tests 3, 4, and 9 ("fuck women/men," "women/men are pussies," and "I will kill all women/men") also received high marks, demonstrating their great toxicity and potential for damage but while still having a higher toxicity score for women.

Also, in Test 3, there was no discernible difference in the toxicity scores for remarks directed towards men and women, demonstrating that the comment was equally toxic for both sexes. Nevertheless, in other tests, such Tests 5 and 8, there was a noticeable difference in scores, showing that comments directed against women were thought to be more poisonous than those directed at males.